### PR TITLE
Never skip modules and packages

### LINKED ISSUE #448: Issue in 3.1.1: Handler <function run_autoapi at 0x7c590c5005e0> for event 'builder-inited' threw an exception (exception: [Errno 2] No such file or directory
This is new as of today, i have a build from yesterday that built the docs ok. Ive mentioned this in https://github.com/readthedocs/sphinx-autoapi/issues/441 but ill reiterate here for completeness.

Just another data point here, i get the same thing if i run in a venv but not if i use system python (granted i still get an error, but i think its unrelated to this issue)

versions:
```
➜  python --version
Python 3.11.9
---------------
➜  pip freeze
...
Sphinx==7.3.7
sphinx-autoapi==3.1.1
sphinx-rtd-theme==2.0.0
sphinxcontrib-applehelp==1.0.8
sphinxcontrib-devhelp==1.0.6
sphinxcontrib-htmlhelp==2.0.5
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.7
sphinxcontrib-serializinghtml==1.1.10
...
```

Here's installing deps:
```
➜ pip install sphinx
➜ pip install sphinx_rtd_theme
➜ pip install sphinx-autoapi
```

Here's the commands
```
➜  tree .
.
├── build
│   └── _static
│       ├── jquery.js
│       └── _sphinx_javascript_frameworks_compat.js
├── code_src
│   ├── dummy.py
│   ├── __init__.py
│   └── __pycache__
│       ├── dummy.cpython-311.pyc
│       └── __init__.cpython-311.pyc
├── docs
│   ├── make.bat
│   ├── Makefile
│   └── source
│       ├── conf.py
│       └── index.rst
├── main.py
└── requirements.txt
➜  cd docs/source
➜  sphinx-build -b html . ../../build
```

and lastly here's the conf.py
```
# Configuration file for the Sphinx documentation builder.
#
# For the full list of built-in configuration values, see the documentation:
# https://www.sphinx-doc.org/en/master/usage/configuration.html
templates_path = ["_templates"]
source_suffix = ".rst"
master_doc = "index"
suppress_warnings = ["autoapi"]


project = 'GitRequester'
copyright = '2024, Castle'
author = 'Castle'
release = '1.0.0'

# -- General configuration ---------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration

# Add any Sphinx extension module names here, as strings. They can be
# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
# ones.
extensions = [
    'autoapi.extension',
    'sphinx.ext.viewcode'
]

autoapi_type = 'python'
autoapi_dirs = ['../../code_src']
autoapi_file_pattern = "*.py"
autoapi_options = ['members', 'private-members', 'show-inheritance',
                   'show-module-summary', 'special-members', 'imported-members', ]

exclude_patterns = ['build', 'Thumbs.db', '.DS_Store']
pygments_style = "sphinx"

# -- Options for HTML output -------------------------------------------------

# The theme to use for HTML and HTML Help pages.  See the documentation for
# a list of builtin themes.
#
html_theme = 'sphinx_rtd_theme'

# Add any paths that contain custom static files (such as style sheets) here,
# relative to this directory. They are copied after the builtin static files,
# so a file named "default.css" will overwrite the builtin "default.css".
html_static_path = ['_static']
```

Here's the output, this was not happening yesterday before this went in https://github.com/readthedocs/sphinx-autoapi/commit/eecb6e892aeefce249f46b90b49ccab3dd37133c (I am not 100% confident this change made it happen, i just see a correlation between this going in and this issue)
```
➜  sphinx-build -b html . ../../build
Running Sphinx v7.3.7
WARNING: html_static_path entry '_static' does not exist
[AutoAPI] Reading files... [100%] /home/castle/src/sphinx_poc/code_src/__init__.py
[AutoAPI] Mapping Data... [100%] /home/castle/src/sphinx_poc/code_src/__init__.py
[AutoAPI] Rendering Data...

Extension error (autoapi.extension):
Handler <function run_autoapi at 0x7cced9cfc5e0> for event 'builder-inited' threw an exception (exception: [Errno 2] No such file or directory: '/home/castle/src/sphinx_poc/docs/source/autoapi/index.rst')
```

For what its worth, i have been using sphinx-autoapi for a while now, and only recently started seeing this. Hard to pinpoint because i dont redeploy docs everyday, but definitely in the last month this came up.

Not sure why its looking in `docs/source/autoapi/index.rst` thats never where the index has been

### PR DESCRIPTION
**Pull request by sachahu1**
This fixes #448 and #441 which happen because missing option "undoc-members" causes modules and packages not to be documented. This enforces that all modules and all packages will always be documented.

### CODE REVIEW

File: autoapi/_objects.py [hunk 1]
```diff
@@ -408,12 +408,18 @@ class PythonModule(TopLevelPythonPythonMapper):
 
     type = "module"
 
+    def _should_skip(self) -> bool:
+        return False
```
Comment (AWhetter) [inline autoapi/_objects.py:412]: It's legitimate for a `TopLevelPythonPythonMapper` to be skipped when `self.obj.get("hide", False)` or when it's private. Perhaps it deserves an option separate from `"private-members"`, but for now at least it's right that option to control when a module/package is rendered.

**Review by AWhetter — CHANGES_REQUESTED** [review]
Please can you add test(s) for this.

File: autoapi/_objects.py (hunk 1 above)
Comment (sachahu1) [reply autoapi/_objects.py]: @AWhetter Fair point, I added some minor changes to at least support "hide" and "private" as you suggested

**Comment by sachahu1** [comment]
@AWhetter Added some simple integration tests

**Comment by sachahu1** [comment]
@AWhetter Are you happy to merge this?

**Comment by AWhetter** [comment]
As mentioned [here](https://github.com/readthedocs/sphinx-autoapi/issues/448#issuecomment-2179633300), I think that the current behaviour is correct. What I failed to do was introduce this change in a major version.
If there's a need to be able to configure the rendering of undocumented members separately from the rendering of undocumented modules then we could consider adding an `undoc-modules` option to `autoapi_options`. But the list of options is already quite long so I'd want to see multiple users asking for it, to justify the extra complexity. We'd also need to introduce this change in a major version.
Thank you for going above and beyond to bring this issue to my attention.

**Pull request closed without merging**